core/trino-main/src/main/java/io/trino/server/security/oauth2/NimbusOAuth2Client.java

```diff
@@ -82,7 +83,7 @@
 import static java.util.concurrent.TimeUnit.SECONDS;
 
 public class NimbusOAuth2Client
-        implements OAuth2Client
+        implements OAuth2Client, LoadableComponent
```

Comment:
@lukasz-walkiewicz Why we didn't use `@PostConstruct` here?

Reply:
IIRC Guice can call `@PostConstruct` multiple times if the method throws an exception.

Reply:
That is good call. Similar issue we experience with validation that we do in guice that is repeated multiple times and then the stack trace from guice is very huge.

Reply:
there are quite a few `@PostConstruct` methods in trino now. 
is guice simply retrying, or is there any special logic behind these methods being called multiple times?
is it configurable?

cc @losipiuk

Reply:
Guice is trying provide a bean multiple times (per each usage?). It is not retry mechanisms, this is more like accidental retry (no backoff etc).

Reply:
https://stackoverflow.com/a/12430864/217852

Reply:
Oh - I think I missed the point of this PR. Is this about the fact that initialization is called multiple times if initialization method throws?
Do we really care about this case enough to make code more complex for that. App will fail to start anyway and I think we can live with longer stacktrace.

Reply:
> Is this about the fact that initialization is called multiple times if initialization method throws?

Yes.

Longer stack trace means sometimes repeating stack trace 200 times.

Not every method likes to be retried and retrying could hide the origin root cause.

Reply:
In case of `OAuth2Client` it connects to the IdP to get it's metadata and if the HTTP call fails it's retried using failsafe policy (retry for up to `http-server.authentication.oauth2.oidc.discovery.timeout` seconds). When guice `@PostConstruct` method is used the timeout can be effectively multiples of the configured value since the method can be invoked more than once.

Reply:
> In case of `OAuth2Client` it connects to the IdP to get it's metadata and if the HTTP call fails

I think no  component should do HTTP calls during server startup.
Consider the IdP server might be temporarily unavailable.
Generally we should assume all network dependencies can be restarted or just started in no particular order.

Can you please remove the network calls from `@PostConstruct` and do them lazy when first needed?